src/main/kotlin/org/neo4j/graphql/handler/filter/OptimizedFilterHandler.kt

```diff
@@ -0,0 +1,369 @@
+            private val current: PropertyContainer<*>,
+            private val variablePrefix: String,
+            private val readingWithoutWhereFactory: () -> StatementBuilder.OngoingReading,
+            private val type: GraphQLFieldsContainer,
+            private val value: Value<*>?,
+            private val filterParams: MutableMap<String, Any?>,
+            private val parentPassThroughWiths: Collection<Expression>
+    ) {
+
+        /**
+         * @param additionalFilter additional filters to be applied to the where
+         */
+        fun parseFilter(additionalFilter: ((StatementBuilder.ExposesWith) -> StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere)? = null): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere {
+            if (value !is ObjectValue) {
+                throw OptimizedQueryException()
+            }
+            // WHERE MATCH all predicates for current
+            // WITH x
+            var query = addWhere(additionalFilter)
+
+            // Handle all quantifier (ALL / NONE / SINGLE / ANY)
+            query = handleRelation(query)
+
+            // Handle OR / AND
+            return handleCombinations(query)
+        }
+
+        private fun addWhere(
+                additionalFilter: ((StatementBuilder.ExposesWith) -> StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere)? = null
+        ): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere {
+            val reading = readingWithoutWhereFactory()
+            if (parsedQuery.fieldPredicates.isEmpty()) {
+                if (reading is StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere) {
+                    return reading
+                }
+            }
+            if (reading !is StatementBuilder.OngoingReadingWithoutWhere) {
+                throw IllegalStateException("cannot handle " + reading::class.java)
+            }
+
+            // WHERE MATCH all predicates for current
+            val readingWithWhere = addConditions<StatementBuilder.OngoingReadingWithWhere>(current, variablePrefix, parsedQuery.fieldPredicates, filterParams) { where, condition ->
+                where?.and(condition) ?: reading.where(condition)
+            }
+            return if (additionalFilter != null) {
+                additionalFilter(readingWithWhere ?: reading)
+            } else {
+                val withs = if (parsedQuery.relationPredicates.isNotEmpty() && parentPassThroughWiths.firstOrNull { it == current.requiredSymbolicName } == null) {
+                    parentPassThroughWiths + current.requiredSymbolicName
+                } else {
+                    parentPassThroughWiths
+                }
+                withClauseWithOptionalDistinct(readingWithWhere ?: reading, withs, useDistinct)
+            }
+        }
+
+        private fun handleRelation(passedQuery: StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere): StatementBuilder.OrderableOngoingReadingAndWithWithoutWhere {
```

Comment:
this function is definitely too long and needs to be broken up into a series of steps